lake-primitives/src/types/block.rs

```diff
@@ -118,128 +92,124 @@ impl Block {
-        }
-        self.events.values().flatten()
+    /// Returns a Vec of [Events](crate::events::Event) emitted in the [Block]
+    pub fn events(&self) -> HashMap<super::ReceiptId, Vec<events::Event>> {
+        self.executed_receipts
+            .iter()
+            .map(|receipt| (receipt.receipt_id(), receipt.events()))
+            .collect()
     }
 
     /// Returns an iterator of the [StateChanges](crate::state_changes::StateChange) happened in the [Block]
-    pub fn state_changes(&mut self) -> impl Iterator<Item = &state_changes::StateChange> {
-        if self.state_changes.is_empty() {
-            self.state_changes = self
-                .streamer_message
-                .shards
-                .iter()
-                .flat_map(|shard| shard.state_changes.iter())
-                .map(Into::into)
-                .collect();
-        }
+    pub fn state_changes(&self) -> impl Iterator<Item = &state_changes::StateChange> {
         self.state_changes.iter()
     }
 
     /// Helper to get all the [Actions](crate::actions::Action) by the single [Receipt](crate::receipts::Receipt)
     ///
     /// **Heads up!** This methods searches for the actions in the current [Block] only.
     pub fn actions_by_receipt_id<'a>(
-        &'a mut self,
+        &'a self,
         receipt_id: &'a super::ReceiptId,
     ) -> impl Iterator<Item = &actions::Action> + 'a {
         self.actions()
             .filter(move |action| &action.receipt_id() == receipt_id)
     }
 
     /// Helper to get all the [Events](crate::events::Event) emitted by the specific [Receipt](crate::receipts::Receipt)
-    pub fn events_by_receipt_id(&mut self, receipt_id: &super::ReceiptId) -> Vec<events::Event> {
-        if self.events.is_empty() {
-            self.build_events_hashmap();
-        }
-        if let Some(events) = self.events.get(receipt_id) {
+    pub fn events_by_receipt_id(&self, receipt_id: &super::ReceiptId) -> Vec<events::Event> {
+        if let Some(events) = self.events().get(receipt_id) {
             events.to_vec()
         } else {
             vec![]
         }
     }
 
     /// Helper to get all the [Events](crate::events::Event) emitted by the specific contract ([AccountId](crate::near_indexer_primitives::types::AccountId))
-    pub fn events_by_contract_id<'a>(
-        &'a mut self,
-        account_id: &'a crate::near_indexer_primitives::types::AccountId,
-    ) -> impl Iterator<Item = &events::Event> + 'a {
+    pub fn events_by_contract_id(
+        &self,
+        account_id: &crate::near_indexer_primitives::types::AccountId,
+    ) -> Vec<events::Event> {
+        let account_id_clone = account_id.clone(); // Clone the account_id
         self.events()
-            .filter(move |event| event.is_emitted_by_contract(&account_id.clone()))
+            .values()
+            .flatten()
+            .filter(|event| event.is_emitted_by_contract(&account_id_clone))
+            .map(Clone::clone)
```

Comment:
Just want to point out that you turned `.events()` to return HashMap only to turn it into an iterator of values and flatten them... Revert the changes here, the old implementation was good (except the unnecessary cloning of `account_id`